### PR TITLE
Config rework

### PR DESCRIPTION
**Pull request by crufter**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -5,23 +5,20 @@ go 1.13
...
 	github.com/go-acme/lego/v3 v3.4.0
+	github.com/gobwas/pool v0.2.1 // indirect
 	github.com/golang/protobuf v1.4.2
 	github.com/google/uuid v1.1.2
 	github.com/gorilla/handlers v1.4.2
 	github.com/gorilla/mux v1.7.3
 	github.com/hashicorp/go-version v1.2.1
 	github.com/juju/fslock v0.0.0-20160525022230-4d5c94c67b4b
-	github.com/micro/go-micro/v3 v3.0.0-beta.2.0.20200918132042-8975184b88a7
+	github.com/micro/go-micro v1.18.0
```
Comment (domwong) [inline go.mod:19]: Why has a v1 import crept in ? need to track this down and fix to v3

Comment (crufter) [reply go.mod:19]: It's VScodde

File: service/config/client/options.go [hunk 2]
```diff
@@ -3,7 +3,7 @@ package client
 import (
 	"context"
 
-	"github.com/micro/go-micro/v3/config/source"
+	"github.com/micro/go-micro/config/source"
```
Comment (domwong) [inline service/config/client/options.go]: should be v3

File: service/config/client/client.go [hunk 3]
```diff
@@ -14,92 +15,67 @@ import (
...
-		return
+		logger.Error(err)
 	}
-	return newWatcher(stream)
 }
 
-// Write is unsupported
-func (m *srv) Write(cs *source.ChangeSet) error {
-	return nil
+func (m *srv) Delete(path string, options ...config.Option) {
```
Comment (domwong) [inline service/config/client/client.go:56]: Think I'm somewhat late to the party here but why don't we return `error` in any of these methods? Just logging error will likely get lost

File: service/config/client/options.go (hunk 2 above)
Comment (crufter) [reply service/config/client/options.go]: Just fixed, this was what caused 1.18, nice catch

File: service/config/client/client.go (hunk 3 above)
Comment (crufter) [reply service/config/client/client.go:56]: I don't know honestly, I kept it as this is how it worked forever but if we change it we should change it now probably.

File: service/config/server/encryption.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+// encrypt/decrypt functions are taken from https://www.melvinvivas.com/how-to-encrypt-and-decrypt-data-using-aes/
+
+func encrypt(stringToEncrypt string, key []byte) (encryptedString string) {
+	plaintext := []byte(stringToEncrypt)
+
+	//Create a new Cipher Block from the key
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		panic(err.Error())
```
Comment (domwong) [inline service/config/server/encryption.go:20]: why panic rather than bubble up an error?

Comment (crufter) [reply service/config/server/encryption.go:20]: copypasta code that's why :trollface:

File: service/config/server/handler.go [hunk 5]
```diff
@@ -23,432 +21,220 @@ const (
...
-			return errors.BadRequest("config.Config.Update", "unmarshal value error: %v", err)
-		}
-		record = records[0]
-	}
-
-	// generate a new base changeset
-	changeSet := &source.ChangeSet{
-		Format: "json",
-		Data:   []byte(`{}`),
+	ch, err := store.Read(req.Namespace)
```
Comment (domwong) [inline service/config/server/handler.go:173]: should be `ns` rather than `req.Namespace`

File: service/config/server/handler.go [hunk 6]
```diff
@@ -23,432 +21,220 @@ const (
...
-
-	if err := store.Write(record); err != nil {
-		return errors.BadRequest("config.Config.Update", "update into db error: %v", err)
 	}
 
-	_ = publish(ctx, &pb.WatchResponse{Namespace: req.Change.Namespace, ChangeSet: req.Change.ChangeSet})
-
-	return nil
+	return store.Write(&store.Record{
+		Key:   req.Namespace,
```
Comment (domwong) [inline service/config/server/handler.go:206]: should be `ns`

File: service/config/server/handler.go [hunk 7]
```diff
@@ -23,432 +21,220 @@ const (
...
-	if len(req.Namespace) == 0 {
-		req.Namespace = defaultNamespace
-	}
-
-	// authorize the request
-	if err := namespace.Authorize(ctx, req.Namespace); err == namespace.ErrForbidden {
-		return errors.Forbidden("config.Config.Watch", err.Error())
-	} else if err == namespace.ErrUnauthorized {
-		return errors.Unauthorized("config.Config.Watch", err.Error())
+	ch, err := store.Read(req.Namespace)
```
Comment (domwong) [inline service/config/server/handler.go:226]: should be ns

File: service/config/server/handler.go [hunk 8]
```diff
@@ -23,432 +21,220 @@ const (
...
-	return reader.Values(ch)
-}
-
-// publish a change
-func publish(ctx context.Context, ch *pb.WatchResponse) error {
-	req := muclient.NewMessage(watchTopic, ch)
-	return muclient.Publish(ctx, req)
+	values.Delete(req.Path)
+	return store.Write(&store.Record{
+		Key:   req.Namespace,
```
Comment (domwong) [inline service/config/server/handler.go:237]: should be `ns`

**Review by domwong — APPROVED** [review]

**Pull request merged**